### PR TITLE
[NeoML] CFunctionWithHessian -- use custom threads

### PR DESCRIPTION
**Pull request by favorart**

### CODE REVIEW

File: NeoML/src/TraditionalML/Function.cpp [hunk 1]
```diff
@@ -17,451 +17,436 @@ limitations under the License.
...
 
-	const int curThreadCount = IsOmpRelevant( vectorCount ) ? threadCount : 1;
+struct CThreadFunctionWithHessianImpl::CProductArgs final : public CThreadFunctionWithHessianImpl::IArgs {
+	CProductArgs( CThreadFunctionWithHessianImpl& f, const CFloatVector& argument );
 
-	CArray<CFloatVector> resultReduction;
-	resultReduction.Add( CFloatVector( result.Size() ), curThreadCount );
+	void Run( int threadIndex, int index, int count ) override;
+	void SubRun( int /*index*/, CFloatVectorDesc, double* /*value*/,
+		CFloatVector& /*grad*/, float /*answer*/, float /*weight*/ ) override {}
```
Comment (FedyuninV) [inline NeoML/src/TraditionalML/Function.cpp:86]: `NeoAssert( false );`

File: NeoML/src/TraditionalML/Function.cpp [hunk 2]
```diff
@@ -17,451 +17,436 @@ limitations under the License.
...
+struct CThreadFunctionWithHessianImpl::IArgs {
+	explicit IArgs( CThreadFunctionWithHessianImpl& f ) :
+		F( f ),
+		ThreadsCount( F.ThreadPool->Size() )
+	{}
+	virtual ~IArgs() {}
+
+	virtual int Size() const { return F.Matrix.Height; }
+	virtual void Run( int threadIndex, int index, int count ) = 0;
+	virtual void SubRun( int index, CFloatVectorDesc,
```
Comment (FedyuninV) [inline NeoML/src/TraditionalML/Function.cpp]: This should be moved to `CSetArgs`.

File: NeoML/src/TraditionalML/Function.cpp [hunk 3]
```diff
@@ -17,451 +17,436 @@ limitations under the License.
...
+		double* value, CFloatVector& grad, float answer, float weight ) = 0;
+	virtual void Reduction() = 0;
+
+	CThreadFunctionWithHessianImpl& F;
```
Comment (FedyuninV) [inline NeoML/src/TraditionalML/Function.cpp]: 1 letter public class field? I agree about variable with a small lifetime within a small function/cycle...

File: NeoML/src/TraditionalML/Function.cpp [hunk 4]
```diff
@@ -17,451 +17,436 @@ limitations under the License.
...
-	errorWeight( static_cast<float>( _errorWeight ) ),
-	l1Coeff( _l1Coeff ),
-	threadCount( _threadCount ),
-	value( 0.f ),
-	answers( data.GetVectorCount() ),
-	weights( data.GetVectorCount() )
+struct CThreadFunctionWithHessianImpl::CSetArgs : public CThreadFunctionWithHessianImpl::IArgs {
+	CSetArgs( CThreadFunctionWithHessianImpl&, const CFloatVector& argument );
+
+	void Run( int threadIndex, int index, int count ) override;
```
Comment (FedyuninV) [inline NeoML/src/TraditionalML/Function.cpp:230]: `final`? You finalize here `Run` method and allow to override `SubRun`

Comment (favorart) [reply NeoML/src/TraditionalML/Function.cpp:230]: fixed

File: NeoML/src/TraditionalML/Function.cpp (hunk 3 above)
Comment (favorart) [reply NeoML/src/TraditionalML/Function.cpp]: fixed

File: NeoML/src/TraditionalML/Function.cpp (hunk 2 above)
Comment (favorart) [reply NeoML/src/TraditionalML/Function.cpp]: fixed

**Review by FedyuninV — APPROVED** [review]

**Pull request merged**